### PR TITLE
Issue 7133 Add timeout option in getThrowingExceptions() 

### LINKED ISSUE #7133: Add Timeout for GetSegmentInfo call.
**Problem description**
Initiate checkpoint call tries to hold lock on RevisionedSTreamClient Object but at the same time other thread which is holding lock on this is stuck at getSegmentInfo call and which is waiting forever to complete the future 

snippet from thread dump

```
"clientInternal-1-1" Id=84 WAITING on java.util.concurrent.CompletableFuture$Signaller@56157aea
	at java.base@11.0.18/jdk.internal.misc.Unsafe.park(Native Method)
	-  waiting on java.util.concurrent.CompletableFuture$Signaller@56157aea
	at java.base@11.0.18/java.util.concurrent.locks.LockSupport.park(LockSupport.java:194)
	at java.base@11.0.18/java.util.concurrent.CompletableFuture$Signaller.block(CompletableFuture.java:1796)
	at java.base@11.0.18/java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3128)
	at java.base@11.0.18/java.util.concurrent.CompletableFuture.waitingGet(CompletableFuture.java:1823)
	at java.base@11.0.18/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1998)
	at io.pravega.common.concurrent.Futures.getThrowingException(Futures.java:184)
	at io.pravega.client.state.impl.RevisionedStreamClientImpl.readFrom(RevisionedStreamClientImpl.java:140)
	-  locked java.lang.Object@21effd0f
	...

	Number of locked synchronizers = 1
	- java.util.concurrent.ThreadPoolExecutor$Worker@6098a473
```

```
"Checkpoint Timer" Id=91 BLOCKED on java.lang.Object@21effd0f owned by "clientInternal-1-1" Id=84
	at io.pravega.client.state.impl.RevisionedStreamClientImpl.writeConditionally(RevisionedStreamClientImpl.java:97)
	-  blocked on java.lang.Object@21effd0f
	at io.pravega.client.state.impl.StateSynchronizerImpl.conditionallyWrite(StateSynchronizerImpl.java:238)
	at io.pravega.client.state.impl.StateSynchronizerImpl.updateState(StateSynchronizerImpl.java:156)
	at io.pravega.client.stream.impl.ReaderGroupImpl.initiateCheckpoint(ReaderGroupImpl.java:157)
	at io.pravega.connectors.flink.ReaderCheckpointHook.triggerCheckpoint(ReaderCheckpointHook.java:116)
	at app//org.apache.flink.runtime.checkpoint.hooks.MasterHooks$WrappedMasterHook.lambda$triggerCheckpoint$1(MasterHooks.java:340)
	at app//org.apache.flink.runtime.checkpoint.hooks.MasterHooks$WrappedMasterHook$$Lambda$1410/0x0000000100c5b840.get(Unknown Source)
	at app//org.apache.flink.util.LambdaUtil.withContextClassLoader(LambdaUtil.java:94)
	...

	Number of locked synchronizers = 1
	- java.util.concurrent.ThreadPoolExecutor$Worker@616e5d45
```

**Problem location**
https://github.com/pravega/pravega/blob/master/client/src/main/java/io/pravega/client/state/impl/RevisionedStreamClientImpl.java#L141C73-L141C87

**Suggestions for an improvement**

Add timeout for SegmentMetadataClient.getSegmentInfo()

### PR DESCRIPTION
**Pull request by a6dulaleem**
**Change log description**
Adding timeout to getThrowingExceptions() and using that method in following spaces  

1. EventStreamReaderImpl
handleSegmentTruncated () it needed because there is some timeout mentioned before calling this method in readNextEventInternal() method.

 
2. RevisionedStreamClientImpl
readFrom() some locking is happening so its mandatory here

Here in this PR we are also adding some timeout inside client config. To take the client config we introduced new constructor without breaking the compatibility in `RevisionStreamClient`.

for now Server timeout is 30sec by default this is configurable and new timeout can be added by creating the client config instance and passing this instance while creating the RevisionStreamClient.

**Purpose of the change**  
to fix #7133 

**What the code does**  

- Add timeout to getThrowingExceptions and use that method.
- Add configuration for timeout in client config
- Introduce new constructor without breaking backward compatibility
- Provide unit test case.

**How to verify it**  
Verifying all the changes through unit test.

### CODE REVIEW

File: common/src/test/java/io/pravega/common/concurrent/FuturesTests.java [hunk 1]
```diff
@@ -77,7 +79,52 @@ public void testGetThrowingExceptions() {
...
+        } catch (InterruptedException e) {
+            assertEquals(true, Thread.interrupted());
+        }
+        assertFalse(Thread.currentThread().isInterrupted());
+
+        assertThrows(TimeoutException.class, () -> getThrowingExceptionWithTimeout(incompleteFuture, 100));
+        // catching TimeoutException
+        try {
+            getThrowingExceptionWithTimeout(incompleteFuture, 100);
+            fail("Expected exception was not thrown.");
```
Comment (tkaitchuck) [inline common/src/test/java/io/pravega/common/concurrent/FuturesTests.java:122]: `AssertExtentions.assertThows` will allow you to write this more neatly.

File: common/src/test/java/io/pravega/common/concurrent/FuturesTests.java [hunk 2]
```diff
@@ -77,7 +79,52 @@ public void testGetThrowingExceptions() {
...
+        CompletableFuture<String> timeoutFuture = new CompletableFuture<String>();
+        assertThrows(TimeoutException.class, () -> getThrowingExceptionWithTimeout(timeoutFuture, 100));
+
+        CompletableFuture<String> incompleteFuture = new CompletableFuture<String>();
+        assertThrows(TimeoutException.class, () -> getThrowingExceptionWithTimeout(incompleteFuture, 100));
+        Thread.currentThread().interrupt();
+        try {
+            getAndHandleExceptions(incompleteFuture, RuntimeException::new);
+            fail();
+            Thread.sleep(1); //Here only to fix compiler error
```
Comment (tkaitchuck) [inline common/src/test/java/io/pravega/common/concurrent/FuturesTests.java:104]: assertThrows should avoid the need for this.

File: common/src/test/java/io/pravega/common/concurrent/FuturesTests.java [hunk 3]
```diff
@@ -77,7 +79,52 @@ public void testGetThrowingExceptions() {
...
+        } catch (InterruptedException e) {
+            assertEquals(true, Thread.interrupted());
+        }
+        Thread.currentThread().interrupt();
+        try {
+            getThrowingExceptionWithTimeout(incompleteFuture, 100);
+            fail();
+            Thread.sleep(1); //Here only to fix compiler error
```
Comment (tkaitchuck) [inline common/src/test/java/io/pravega/common/concurrent/FuturesTests.java:112]: same here.

File: client/src/main/java/io/pravega/client/ClientConfig.java [hunk 4]
```diff
@@ -103,6 +105,10 @@ public class ClientConfig implements Serializable {
      */
     private final int maxConnectionsPerSegmentStore;
 
+    // This is a timeout to be passed to Completeablefuture to complete.
+    private final long timeoutToCompleteFutureMilliSec;
```
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/ClientConfig.java:109]: Name this better. (It's not about "completeFuture")

File: client/src/main/java/io/pravega/client/ClientConfig.java [hunk 5]
```diff
@@ -45,6 +46,7 @@
 public class ClientConfig implements Serializable {
 
     static final int DEFAULT_MAX_CONNECTIONS_PER_SEGMENT_STORE = 10;
+    static final long READ_TIMEOUT_MS = TimeUnit.SECONDS.toMillis(30);
```
[the remark this answers is not in the document]
Comment (a6dulaleem) [reply client/src/main/java/io/pravega/client/ClientConfig.java]: done

File: client/src/main/java/io/pravega/client/stream/impl/EventStreamReaderImpl.java [hunk 6]
```diff
@@ -133,11 +136,13 @@ public EventRead<Type> readNextEvent(long timeoutMillis) throws Reinitialization
             } catch (ReaderNotInReaderGroupException e) {
                 close();
                 throw new ReinitializationRequiredException(e);
+            } catch (TimeoutException e) {
+                throw new ServerTimeoutException(format("Event is not read {%s}", e));
```
[the remark this answers is not in the document]
Comment (a6dulaleem) [reply client/src/main/java/io/pravega/client/stream/impl/EventStreamReaderImpl.java]: now i am not throwing server timeout exception i am just catching and logging a warning.

File: client/src/main/java/io/pravega/client/stream/impl/EventStreamReaderImpl.java [hunk 7]
```diff
@@ -170,7 +175,7 @@ private EventRead<Type> readNextEventInternal(long timeoutMillis) throws ReaderN
                     handleEndOfSegment(segmentReader, isSegmentSealed);
                     buffer = null;
                 } catch (SegmentTruncatedException e) {
-                    handleSegmentTruncated(segmentReader);
+                    handleSegmentTruncated(segmentReader, firstByteTimeoutMillis);
```
[the remark this answers is not in the document]
Comment (a6dulaleem) [reply client/src/main/java/io/pravega/client/stream/impl/EventStreamReaderImpl.java]: now passing timeoutMillis

File: client/src/main/java/io/pravega/client/stream/impl/EventStreamReaderImpl.java [hunk 8]
```diff
@@ -395,15 +400,15 @@ private void handleEndOfSegment(EventSegmentReader oldSegment, boolean segmentSe
...
+    private void handleSegmentTruncated(EventSegmentReader segmentReader, long timeoutInMilli) throws TruncatedDataException, TimeoutException {
         Segment segmentId = segmentReader.getSegmentId();
         log.info("{} encountered truncation for segment {} ", this, segmentId);
 
         @Cleanup
         SegmentMetadataClient metadataClient = metadataClientFactory.createSegmentMetadataClient(segmentId,
                 DelegationTokenProviderFactory.create(controller, segmentId, AccessOperation.READ));
         try {
-            long startingOffset = Futures.getThrowingException(metadataClient.getSegmentInfo()).getStartingOffset();
+            long startingOffset = Futures.getThrowingExceptionWithTimeout(metadataClient.getSegmentInfo(), timeoutInMilli).getStartingOffset();
```
[the remark this answers is not in the document]
Comment (a6dulaleem) [reply client/src/main/java/io/pravega/client/stream/impl/EventStreamReaderImpl.java:407]: yeah so now I am catching it here and logging warning

File: client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java [hunk 9]
```diff
@@ -498,11 +517,15 @@ public void write(PendingEvent event) {
             try {
                 // if connection is null getConnection() establishes a connection and retransmits all events in inflight
                 // list.
-                connection = Futures.getThrowingException(getConnection());
+                connection = Futures.getThrowingExceptionWithTimeout(getConnection(), clientConfig.getConnectTimeoutMilliSec());
```
Comment (tkaitchuck) [inline client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java:520]: So, looking at this there already is a timeout of sorts on line 248. The problem is the internal thread can be blocked on the getInfo without time. This means the future on 673 would never return. So if this timeout was introduced all requests would simply timeout here and the internal thread would never unblock and the reader would never work.

This seemed familiar, and I looked, and in fact we made this change before: https://github.com/pravega/pravega/pull/6528/files#diff-77e930ef537f025820cd9af3e14f393bef92097b9f7978e7f3b57bfc3f28a1ab
That change ended up being rolled back due to connection churn, but I think it's approach to this file is the right approach. For now, let's copy that approach to this file except the change to the `reconnect` call below, for that one I think we'll need to have a meeting to discuss, as `reconnect` is definitely not right, but it did cause problems to change it.

File: client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java [hunk 10]
```diff
@@ -718,4 +718,4 @@ public List<PendingEvent> getUnackedEventsOnSeal() {
     public long getLastObservedWriteOffset() {
         return state.getLastSegmentLength();
     }
-}
+}
```
Comment (a6dulaleem) [inline client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java:721]: Please ignore this, tried multiple time, I am not getting any diff with master in this file.

File: client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java (hunk 9 above)
Comment (a6dulaleem) [reply client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java]: will do this in #7194

**Review by tkaitchuck — APPROVED** [review]

**Pull request merged**